### ISSUE TITLE
PROTOCOL_ERROR (0x1): frame_header: not enough input

### ISSUE THREAD

**Issue by kit-ty-kate** [roburio/http-lwt-client#7]
The following code:
```
let () =
  Lwt_main.run begin
    match%lwt
      Http_lwt_client.one_request
        ~meth:`POST
        ~headers:[("Content-type", "application/json")]
        ~body:{|{"username": "opam-health-check", "text":"The latest check is done. Check out http://check.ocamllabs.io/diff/1639737025-87bf0c8a02e89350c38353a95c84b0863f710e49..1639928420-997305af092a2f2d893c0f1c4ed59ee8c2a2f342 to discover which packages are now broken or fixed"}|}
        "<my test slack webhook>"
    with
    | Ok ({Http_lwt_client.status = `OK; _}, _body) -> Lwt.return_unit
    | Ok (resp, body) ->
        let resp = Format.asprintf "%a" Http_lwt_client.pp_response resp in
        let body = match body with None -> "" | Some body -> "\nBody: "^body in
        prerr_endline ("Webhook returned failure: "^resp^body);
        Lwt.return_unit
    | Error (`Msg msg) -> prerr_endline ("Webhook failed with: "^msg); Lwt.return_unit
  end
```
fail with:
```
PROTOCOL_ERROR (0x1): frame_header: not enough input
```
With advices from @dinosaure I changed the default HTTP config from HTTP2 to HTTP1.1 (using ``~config:(`HTTP_1_1 Httpaf.Config.default)``) and it worked fine.

Something seems to be broken by default in H2.

**Comment by hannesm** [comment]
Thanks for your report, @kit-ty-kate. I wonder whether this is reproducible with https://github.com/anmonteiro/piaf? If it is not reproducible, it makes sense to look into the http-lwt-client usage of h2; if reproducible, it is likely an issue in h2 and should be reported there.

I remember a similar issue was around, but fixed with https://github.com/roburio/http-lwt-client/commit/b01c51206f2104e7a714f5025c52ff9339213635 -- do you use the latest release (0.0.5)?

**Comment by hannesm** [comment]
Is there a (public) url (preferable GET) to reproduce the issue? I tried with https://slack.com which works fine with `hurl.exe`.

**Comment by kit-ty-kate** [comment]
it is reproducible with a slack webhook url. You can create one by following these steps: https://api.slack.com/messaging/webhooks

If it's too much hassle I can send a webhook privately

**Comment by hannesm** [comment]
with more recent changes in h2 (and further work in the http-lwt-client), does the problem persist? if so, it would indeed be great if you could provide an endpoint that fails this way (of course via a private message - matrix, email).

**Comment by ScoreUnder** [comment]
I've been trying to use this library to send a POST request and I can't seem to get it to work - GET works fine, but any POST fails. I get a simple "400" from nginx, but I get a similar message, "PROTOCOL_ERROR (0x1): ", when using `https://httpbin.org/post`.

Echoing that it also works fine with HTTP/1.1.

`h2                      0.10.0`
`http-lwt-client         0.2.3`

~~i'm going to try piaf, though i'll have to learn it first!~~ ~~Looks like piaf also doesn't POST correctly~~ discussed with the piaf guy and it can in fact post correctly, I just mixed up a couple of parameters

**Comment by dinosaure** [comment]
> discussed with the piaf guy and it can in fact post correctly, I just mixed up a couple of parameters

Which parameters did you mix? We probably can improve the documentation and/or the API to avoid such issue.

**Comment by ScoreUnder** [comment]
@dinosaure I had searched for "piaf post ocaml" to find the docs and reached here: <https://docs.ocaml.pro/docs/LIBRARY.piaf@piaf.0.1.0/Piaf/Client/index.html>

I saw

> A client instance represents a connection to a single remote endpoint, and the remaining functions in this module will issue requests to that endpoint only.

So I put the full URL to the endpoint in the `Uri.t` parameter, and was then confused to see an "extra" `string` parameter on each of the other methods. I assumed it was simply the 'real' URL to make alongside the request and that the `create` must have been using the Uri.t to make the initial connection (for TLS SNI maybe), which turned out to be the wrong assumption and both URLs got concatenated, confusing the server.

**Comment by dinosaure** [comment]
I can't quite understand your situation. Do you use `piaf` or `http-lwt-client`? In the first case, you should postpone an issue on the official [piaf](https://github.com/anmonteiro/piaf/) repository in order to clarify the documentation.

**Comment by ScoreUnder** [comment]
@dinosaure Both! I have been trying to select a library that can do HTTP/2.0 POST requests with a simple interface and those two were both in the running for experimentation.

To clear up my statement way above, I managed to get piaf functioning correctly, but I did not manage to get http-lwt-client working in the same situation. To recreate the code that did not work in http-lwt-client:

```ocaml
utop # Http_lwt_client.request ~meth:`POST ~headers:["Content-Type","application/json"] ~body:"{\"test\":123}" "https://httpbin.org/post" (fun _ a b -> Lwt.return @@ a ^ b) "" |> Lwt_main.run;;
- : (Http_lwt_client.response * string, [> `Msg of string ]) result =
Error (`Msg "PROTOCOL_ERROR (0x1): ")
```

**Comment by hannesm** [comment]
Thanks for your example. Maybe the http-lwt-client should lowercase the header keys by default...

To get your example working, the `"content-type"` must be all lowercase (you had `"Content-Type"`):

```OCaml
# Http_lwt_client.request ~meth:`POST ~headers:["content-type","application/json"] ~body:"{\"test\":123}" "https://httpbin.org/post" (fun _ a b -> Lwt.return @@ a ^ b) "" |> Lwt_main.run;;
- : (Http_lwt_client.response * string, [> `Msg of string ]) result =
Ok
 ({Http_lwt_client.version = {Http_lwt_client.Version.major = 2; minor = 0};
   status = `OK; reason = ""; headers = <abstr>},
  "{\n  \"args\": {}, \n  \"data\": \"{\\\"test\\\":123}\", \n  \"files\": {}, \n  \"form\": {}, \n  \"headers\": {\n    \"Content-Length\": \"12\", \n    \"Content-Type\": \"application/json\", \n    \"Host\": \"httpbin.org\", \n    \"X-Amzn-Trace-Id\": \"Root=1-64a290c7-6a6832524428f6c42a320a2d\"\n  }, \n  \"json\": {\n    \"test\": 12"... (* string length 362; truncated *))
```

**Comment by ScoreUnder** [comment]
Interesting! I note that the original code in this thread has that same issue. All my life I've dealt with the "canonical" casing being title case, ~~did that become taboo with HTTP/2?~~ Apparently yes[\[1\]](https://stackoverflow.com/a/41169947/1091693) [\[2\]](https://www.rfc-editor.org/rfc/rfc7540#section-8.1.2)

> However, header field names MUST be converted to lowercase prior to their encoding in HTTP/2. A request or response containing uppercase header field names MUST be treated as malformed.

**Comment by hannesm** [comment]
Indeed, HTTP/2 requires lowercase.... since there have been multiple reports, I'd keep this open until we have a "lowercase headers" in place for http/2.

**Comment by hannesm** [comment]
fixed by #19, released as 0.2.4
